### ISSUE TITLE
custom-elements-es5-adapter.js hilariously uses es6 code

### ISSUE THREAD

**Issue by jonmann20** [webcomponents/webcomponentsjs#749]
Somewhat ironically https://github.com/webcomponents/webcomponentsjs/blob/master/custom-elements-es5-adapter.js is in es6, so when trying to use it to polyfill IE 11, you get an error.

**Comment by MartinsThiago** [comment]
This is expected to happen, es6 adapter target is to prevent errors is browser that support the v1 spec... because you cannot convert your elements to es5 and run on latest browsers

It's not so comfortable having error being thrown, but anyway... it works :)

**Comment by rictic** [comment]
Yep, as @MartinsThiago says, this is necessary when shipping ES5 to a browser that supports native custom elements, as native elements are ES6 classes and when extending an ES6 class you _must_ call `super()`, and there's no ES5 way to do that.

The syntax error being thrown in IE11 and friends is an issue though, and we're tackling it in `polymer build` here: https://github.com/Polymer/polymer-cli/issues/627

**Comment by MartinsThiago** [comment]
Maybe we should consider renaming this file
es5-adapter... may lead people to think it's a polyfill to older browsers

_Naming suggestion: custom-elements-run-your-es5-compiled-element-on-modern-browsers.js_

**Comment by jonmann20** [comment]
Ah I see, ok thank you for clarifying!

**Comment by CITguy** [comment]
I'm still confused. How do I use the adapter if it's meant for ES5 browsers that don't support ES6, but it's written with ES6 syntax?

**Comment by valdrinkoshi** [comment]
The ES5 adapter is needed for native Custom Elements.
Native Custom Elements must be defined with a class (ES6), but most projects transpile to ES5 for reasons (e.g. better tooling available for minification/bundling, or support older browsers).
When you transpile to ES5, you need an adapter for native Custom Elements to work (e.g. in Chrome or Safari).
When you load the adapter in a browser that doesn't support ES6, it will fire a syntax error, and that's expected to happen, because you cannot feature detect the support of ES6 at run time.

In short: the ES5 adapter is written in ES6 to enable native Custom Elements to be declared in ES5 syntax

**Comment by rictic** [comment]
I think we should rename the shim one last time, as this is a repeated point of confusion.

Serious proposal: `es6-shim-for-es5-in-ce-browsers.js`

**Comment by jehna** [comment]
How about `customelements-fix-es5-new-htmlelement-error-in-ce-browsers.js`?

**Comment by svastanesto** [comment]
Hi @rictic! You addressed the problem with syntax error in Polymer case, but what about projects that don't use Polymer? 

My karma test flow breaks in IE 11, due to that error, and I cannot include the script through workaround wrapper code since I can control the markup of the page where tests run (I can only pass the scripts I want included).

**Comment by northkode** [comment]
This is unfortunate as webpack Uglify fails on es6 syntax. 
My current system works great for webcomponents with webpack but this breaks it all.

**Comment by justinfagnani** [comment]
Don't pass the adapter (or other polyfills) to webpack, just load it in a script tag, and only in browsers with native Custom Elements support, currently the latest Safari and Chrome.

**Comment by northkode** [comment]
@justinfagnani how does this work then for ie11 ?  my webpack converts all es6 to es2015 which is why i use the adapter. 

IE 11 completely breaks

**Comment by justinfagnani** [comment]
@northkode you don't need the adapter in IE11, because it doesn't support native Custom Elements.

**Comment by justinfagnani** [comment]
We have a snippet which will make sure that the adapter doesn't load on browsers that don't need it, for when you can't control the HTML that's generated. You need to put this in your HTML before the Custom Element's polyfill:

```html
<script>if (!window.customElements) { document.write('<!--'); }</script>
<script type="text/javascript" src="../webcomponentsjs/custom-elements-es5-adapter.js"></script>
<!--! do not remove -->
```

**Comment by northkode** [comment]
@justinfagnani awesome, thanks.

**Comment by owencm** [comment]
Note document.write is a function browsers recommend you avoid because it is know to create performance issue.

I’d recommend you look instead at the solution described by Rob Dodson of the Chrome team here: https://youtu.be/Ucq9F-7Xp8I?t=7m57s

**Comment by fanyer** [comment]
it should be renamed . And for the best, it should add a notification that 'this adapter is designed for v0 elements run in moden browser '.
the adapter itself actually cannot run in es5. and it is not like babel which can transclude es6 to es5

**Comment by robdodson** [comment]
If you're transpiling with Babel it might be easier to include https://github.com/github/babel-plugin-transform-custom-element-classes. I'm curious to hear how folks get along with it.

**Comment by mcv** [comment]
This whole issue is very confusing. If this is not necessary to get custom-elements working in IE11, then what is?

**Comment by web-padawan** [comment]
@mcv this is needed to get native Custom Elements work after transpiling to ES5. I am wondering if this [README section](https://github.com/webcomponents/webcomponentsjs#custom-elements-es5-adapterjs) is enough to understand?

**Comment by mcv** [comment]
I think I now get it, but it's very confusing. Custom elements already work natively in some browsers (like Chrome) and don't need anything else to work as long as they are in ES6. IE11 supports neither custom elements nor ES6, so you need to compile down to ES5 and add webcomponents-lite.js. But Chrome ignores webcomponents-lite.js and doesn't accept webcomponents in ES5, so it needs this file to get it working again. But then this file breaks IE11.

So if I understand correctly, there's no way to get the same code working in both browsers. If you compile down to ES5, Chrome needs this, and IE11 needs it to be absent. If you can serve different files to different browsers, it would be best to serve Chrome ES6 and serve IE11 the downcompiled version with the polyfill.

Is that correct?

**Comment by justinfagnani** [comment]
@mcv, that's right. It's unfortunately just a confusing set of constraints. I've tried to make it clear, but it's difficult.

Note that there's some hope for simplifying this with a recent change to Babel that allows extending built-in classes like `HTMLElement` by using `Reflect.construct` so emulate a `super()` call. A previous version of this shim used `Reflect.construct` and it was really slow, so I changed it to the current wrapper implementation. It could be that VMs have gotten much better with `Reflect.construct` in the meantime though, making it viable. We'll do some testing as soon as the Babel change works with mixins.

**Comment by Dmitri-Sintsov** [comment]
Isn't it a bit too early to have ES6-only custom elements? Was it so much necessary to deprecate document.registerElement()? IE11 is still around for few years. ES5 is fine language not much worse than ES6, it should be very much capable to define custom elements logic. I do not like transpiling and try to avoid it as much as possible, but someone always tries to enforce me to do that.

**Comment by rictic** [comment]
document.registerElement will never be implemented in any browser but Chrome. The point of web components is to have a web-native component system.

There's a reasonable argument to be made that an ES5 constructor function should be able to extend HTMLElement, but that's not how the specs shook out. There was a debate, and we don't all love the outcome, but at this point it's well settled territory and there's no point continuing to litigate it.

There are a lot of really nice features that make transpilation 100% worth it for me. Native classes are fine, but async/await is _revolutionary_.
